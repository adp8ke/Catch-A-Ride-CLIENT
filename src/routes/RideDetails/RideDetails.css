.ride-details{
  margin-bottom: 2em;
  font-size: 1.1em;
  line-height: 1.5;
  border:1px solid white;
  width:90%;
  background:rgb(28, 35, 64,.67);
  margin:0 auto;
  border-radius: 5px;
}

#ride-btn button{
  padding: 1em;
  margin: .5em;
  border: 1px solid #b8e3ea;
  border-radius:3px;
  background: #b8e3ea;
  color: black;
  min-height: 31px;
  transition: 1.3s;
  margin-top: 5px;
}

#ride-btn button:hover{
  cursor: pointer;
  border: 3px solid #F2DE6A;
  background: #F2DE6A;
  color: black;
  min-height: 35px;
  cursor: pointer;
}

div {
  position: relative;
}

.messageBox, .errorBox {
  font-size: 18px;
  padding: 5px 7px;
  width: 90%;
  margin: 0 auto;
}

.messageBox {
  border: 1px solid white;
  border-radius: 5px;
  margin-bottom:10px;
  background: rgb(28, 35, 64,.67);
}

.messageButton {
  background-color: #F2F9FB;
  border: 1px solid black;
  border-radius: 5px;
}
.messageButton:hover{
  cursor: pointer;
}

.errorBox {
  color: #FC5C66;
  border: 1px solid #FC5C66;
  border-radius: 5px;
  margin-bottom: 10px;
  background: rgb(0, 0, 0, .63);
}

.errorButton {
  color: #FC5C66;
  border: 1px solid #FC5C66;
  background-color: #F2F9FB;
  border-radius: 5px;
}

.errorButton:hover{
  cursor: pointer;
}

.messageButton, .errorButton {
  margin-left: 10px;
  vertical-align: middle;
}
#ride-btn{
  margin-bottom: 10px;
}

@media only screen and (min-width: 700px) {
  .messageBox {
    width: 30%;
    font-size: 20px;
  }
  .ride-details{
    width:68%;
  }

  .errorBox {
    width: 40%;
    font-size: 20px;
  }
}

@media only screen and (min-width:1000px){
  .ride-details{
    width:40%;
  }
}