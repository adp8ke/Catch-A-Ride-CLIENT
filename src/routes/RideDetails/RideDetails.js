import React, { Component } from 'react';
import config from '../../config';
import RideContext from '../../context/RideContext';
import Gmaps from '../../components/Maps/Gmaps';
import TokenService from '../../services/token-service';
import PassengerApiService from '../../services/RidesService/rides-passenger-service';
import DriverApiService from '../../services/RidesService/rides-driver-service';
import './RideDetails.css';
import EditModal from '../../components/EditModal/EditModal';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faMapMarkedAlt, faTrashAlt, faUserSlash, faEdit } from '@fortawesome/free-solid-svg-icons';


export default class RideDetails extends Component {
  static contextType = RideContext;
  static defaultProps = {
    userContext: {
      user: { user_id: '' }
    }
  }

  constructor(props) {
    super(props);
    this.state = {
      error: null,
      message: null,
      isEditing: false,
      remainingSeats: 0,
    };
  }
  componentDidMount() {
    fetch(`${config.API_ENDPOINT}/rides/${this.props.match.params.ride_id}`, {
      method: 'GET',
      headers: {
        'content-type': 'application/json',
        Authorization: `bearer ${TokenService.getAuthToken()}`
      }
    })
      .then(res =>
        (!res.ok)
          ? res.json().then(e => Promise.reject(e))
          : res.json()
      )
      .then(data => this.context.setRide(data))
      .then(() => {
        let startLat = this.context.ride.startCoorLat;
        let startLng = this.context.ride.startCoorLong;
        this.context.setStartingC(startLat, startLng);
        let destLat = this.context.ride.destCoorLat;
        let destLng = this.context.ride.destCoorLong;
        this.context.setDestinationC(destLat, destLng);
        let remainingSeats = this.setRemainingSeats();
        this.setState({ remainingSeats });
      })
      .catch(res => this.setState({ error: res.error },
        () => {
          if (this.state.error === 'unauthorized request') this.props.userContext.setLoggedOut();
        }));
  }

  handleJoin = (ride_id) => {
    PassengerApiService.passengerJoinRide(ride_id)
      .then(res => {
        this.context.setRide(res);
        let remainingSeats = this.setRemainingSeats();
        this.setState({ message: 'You have joined this ride', remainingSeats });
      })
      .catch(res => this.setState({ error: res.error },
        () => {
          if (this.state.error === 'unauthorized request') this.props.userContext.setLoggedOut();
        }));
  }

  handleCancel = (ride_id) => {
    PassengerApiService.passengerCancelRide(ride_id)
      .then(res => {
        this.context.setRide(res);
        let remainingSeats = this.setRemainingSeats();
        this.setState({ message: 'You have left this ride', remainingSeats });
      })
      .catch(res => this.setState({ error: res.error },
        () => {
          if (this.state.error === 'unauthorized request') this.props.userContext.setLoggedOut();
        }));
  }

  handleDelete = (ride_id) => {
    this.context.deleteRide(ride_id);

    return DriverApiService.deleteRide(ride_id)
      .catch(res => (res.error)
        ? this.setState({
          error: res.error
        },
        () => {
          if (this.state.error === 'unauthorized request') this.props.userContext.setLoggedOut();
        })
        : this.props.history.push('/rides')
      );
  }

  handleErrorClose = () => {
    this.setState({ error: null });
  }

  handleMessageClose = () => {
    this.setState({ message: null });

  }
  createEditForm = () => {
    this.setState({ isEditing: true });
  }

  closeEditForm = () => {
    this.setState({ isEditing: false });
  }

  handleEditForm = () => {
    let ride_id = this.context.ride.id;
    let description = document.getElementById('newDescription').value;
    let starting = document.getElementById('newStarting').value;
    let destination = document.getElementById('newDestination').value;
    let date = document.getElementById('newDate').value;
    let time = document.getElementById('newTime').value;

    let updatedDetails = { starting, destination, description, date, time };
    DriverApiService.editRideDetails(ride_id, updatedDetails)
      .then(res => {
        this.context.setRide(res);

        let startLat = this.context.ride.startCoorLat;
        let startLng = this.context.ride.startCoorLong;
        this.context.setStartingC(startLat, startLng);
        let destLat = this.context.ride.destCoorLat;
        let destLng = this.context.ride.destCoorLong;
        this.context.setDestinationC(destLat, destLng);
        this.setState({ isEditing: false });
      })
      .catch(res => this.setState({ error: res.error },
        () => {
          if (this.state.error === 'unauthorized request') this.props.userContext.setLoggedOut();
        }));
  }

  setRemainingSeats = () => {
    let remainingSeats = 0;
    let count = 0;

    let sRArray = Object.keys(this.context.ride);

    for (let i = 6; i < this.context.ride.capacity + 6; i++) {
      count++;
      if (this.context.ride.capacity < count) {
        break;
      }

      if (this.context.ride[sRArray[i]] === null) {
        remainingSeats++;
      }
    }

    if (remainingSeats === 0) {
      return remainingSeats = 'This ride is full';
    }
    else {
      return remainingSeats;
    }
  }

  render() {
    const { error, message, remainingSeats } = this.state;
    const { id, starting, destination, date_time, driver_name, } = this.context.ride;
    let dateStr = new Date(date_time).toLocaleString(undefined, { timeZone: 'UTC' }); let newStr = dateStr.split(', ');
    let dateFormat = newStr[0];
    let time = newStr[1];
    let timeFormat = '';
    if (!time) {
      timeFormat = 'Invalid Date';
    } else {
      let timeArr = time.split(':');
      let amPM = timeArr[2].split(' ');
      timeFormat = `${timeArr[0]}:${timeArr[1]} ${amPM[1]}`;

    }

    const { user_id } = this.props.userContext.user;

    return this.context.ride
      ? <>
        <h2>Ride Details</h2>
        <div className="google-map">
          <Gmaps />
        </div>
        {message && <div className='messageBox'>{message}<button className='messageButton' aria-label='close' onClick={() => this.handleMessageClose()}>X</button></div>}
        {error && <div className='errorBox'>{error}<button className='errorButton' aria-label='close' onClick={() => this.handleErrorClose()}>X</button></div>}
        <div className='ride-details'>
          <p>Driver: {driver_name}</p>
          <p>Meetup Address: {starting}</p>
          <p>Destination: {destination}</p>
          <p>Meetup Date: {dateFormat}</p>
          <p>Meetup Time: {timeFormat}</p>
          <p>Remaining Seats: {remainingSeats}</p>
          <h4>Ride Description:</h4>
          <p>{this.context.ride.description}</p>
          <div id="ride-btn">
            {this.context.ride.driver_id === user_id
              ? <>
                <button type="button" onClick={() => this.handleDelete(id)}>Delete Ride <FontAwesomeIcon icon={faTrashAlt} /></button>
                <button type="button" onClick={() => this.createEditForm()}>Edit Details <FontAwesomeIcon icon={faEdit} /></button>
              </>
              : <>
                <button type="button" onClick={() => this.handleJoin(id)}>Join <FontAwesomeIcon icon={faMapMarkedAlt} /></button>
                <button type="button" onClick={() => this.handleCancel(id)}>Cancel Ride <FontAwesomeIcon icon={faUserSlash} /></button>
              </>}
            {this.state.isEditing && <EditModal handleEditForm={this.handleEditForm} closeEditForm={this.closeEditForm} timeFormat={timeFormat} dateFormat={dateFormat} />}
          </div>
        </div>
      </>
      : <div>Loading</div>;
  }
}
