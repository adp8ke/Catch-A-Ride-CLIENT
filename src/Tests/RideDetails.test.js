import React from 'react';
import ReactDOM from 'react-dom';
import {MemoryRouter} from 'react-router-dom';
import RideDetails from '../routes/RideDetails/RideDetails';

it('renders without crashing', () => {
  const div = document.createElement('div');
  const match = {path: '/rides/:ride_id', url: '/rides/cd684ce0-e001-41df-aa43-af02536a7519', isExact: true, params: {ride_id: 'cd684ce0-e001-41df-aa43-af02536a7519'}};
  ReactDOM.render(<MemoryRouter><RideDetails match={match}/></MemoryRouter>, div);
  ReactDOM.unmountComponentAtNode(div);
});
