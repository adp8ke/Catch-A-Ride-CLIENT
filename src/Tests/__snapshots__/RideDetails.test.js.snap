// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the UI as expected 1`] = `
Array [
  <h2>
    Ride Details
  </h2>,
  <div
    className="google-map"
  >
    <div>
      Loading...
    </div>
  </div>,
  <div
    className="ride-details"
  >
    <p>
      Driver: 
    </p>
    <p>
      Meetup Address: 
    </p>
    <p>
      Destination: 
    </p>
    <p>
      Meetup Date: 
      Invalid Date
    </p>
    <p>
      Meetup Time: 
      Invalid Date
    </p>
    <p>
      Remaining Seats: 
      0
    </p>
    <h4>
      Ride Description:
    </h4>
    <p />
    <div
      id="ride-btn"
    >
      <button
        onClick={[Function]}
        type="button"
      >
        Join 
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-map-marked-alt fa-w-18 "
          data-icon="map-marked-alt"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 576 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M288 0c-69.59 0-126 56.41-126 126 0 56.26 82.35 158.8 113.9 196.02 6.39 7.54 17.82 7.54 24.2 0C331.65 284.8 414 182.26 414 126 414 56.41 357.59 0 288 0zm0 168c-23.2 0-42-18.8-42-42s18.8-42 42-42 42 18.8 42 42-18.8 42-42 42zM20.12 215.95A32.006 32.006 0 0 0 0 245.66v250.32c0 11.32 11.43 19.06 21.94 14.86L160 448V214.92c-8.84-15.98-16.07-31.54-21.25-46.42L20.12 215.95zM288 359.67c-14.07 0-27.38-6.18-36.51-16.96-19.66-23.2-40.57-49.62-59.49-76.72v182l192 64V266c-18.92 27.09-39.82 53.52-59.49 76.72-9.13 10.77-22.44 16.95-36.51 16.95zm266.06-198.51L416 224v288l139.88-55.95A31.996 31.996 0 0 0 576 426.34V176.02c0-11.32-11.43-19.06-21.94-14.86z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
      <button
        onClick={[Function]}
        type="button"
      >
        Cancel Ride 
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-user-slash fa-w-20 "
          data-icon="user-slash"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 640 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M633.8 458.1L362.3 248.3C412.1 230.7 448 183.8 448 128 448 57.3 390.7 0 320 0c-67.1 0-121.5 51.8-126.9 117.4L45.5 3.4C38.5-2 28.5-.8 23 6.2L3.4 31.4c-5.4 7-4.2 17 2.8 22.4l588.4 454.7c7 5.4 17 4.2 22.5-2.8l19.6-25.3c5.4-6.8 4.1-16.9-2.9-22.3zM96 422.4V464c0 26.5 21.5 48 48 48h350.2L207.4 290.3C144.2 301.3 96 356 96 422.4z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </div>
  </div>,
]
`;
